File: Robotics Kinematics and Dynamics Course/Chapter10/Lesson2/Chapter10_Lesson2.cpp
#include <iostream>
#include <Eigen/Dense>

using Eigen::Matrix2d;
using Eigen::Vector2d;

Matrix2d jacobian2R(double q1, double q2, double l1, double l2) {
    double s1  = std::sin(q1);
    double c1  = std::cos(q1);
    double s12 = std::sin(q1 + q2);
    double c12 = std::cos(q1 + q2);

    Matrix2d J;
    J(0,0) = -l1 * s1 - l2 * s12;
    J(0,1) = -l2 * s12;
    J(1,0) =  l1 * c1 + l2 * c12;
    J(1,1) =  l2 * c12;
    return J;
}

Vector2d generalizedTorquesFromForce(double q1, double q2,
                                     double l1, double l2,
                                     const Vector2d &f) {
    Matrix2d J = jacobian2R(q1, q2, l1, l2);
    return J.transpose() * f;
}

int main() {
    double l1 = 1.0, l2 = 0.8;
    double q1 = 0.5, q2 = -0.3;
    Vector2d f;
    f << 10.0, 5.0;

    Vector2d tau = generalizedTorquesFromForce(q1, q2, l1, l2, f);
    std::cout << "Joint torques: " << tau.transpose() << std::endl;

    return 0;
}
      

File: Robotics Kinematics and Dynamics Course/Chapter10/Lesson2/Chapter10_Lesson2.py
import numpy as np

def jacobian_2r(q, l1, l2):
    q1, q2 = q
    s1 = np.sin(q1)
    c1 = np.cos(q1)
    s12 = np.sin(q1 + q2)
    c12 = np.cos(q1 + q2)

    J11 = -l1 * s1 - l2 * s12
    J12 = -l2 * s12
    J21 =  l1 * c1 + l2 * c12
    J22 =  l2 * c12
    J = np.array([[J11, J12],
                  [J21, J22]])
    return J

def generalized_torques_from_force(q, l1, l2, f):
    """
    q : array-like, shape (2,)
    f : array-like, shape (2,) planar end-effector force [fx, fy]
    """
    J = jacobian_2r(q, l1, l2)
    f_vec = np.asarray(f).reshape(2,)
    tau = J.T @ f_vec
    return tau

if __name__ == "__main__":
    l1, l2 = 1.0, 0.8
    q  = np.array([0.5, -0.3])  # radians
    f  = np.array([10.0, 5.0])  # N in x-y plane

    tau = generalized_torques_from_force(q, l1, l2, f)
    print("Joint torques:", tau)

    # Numerical check of virtual work invariance:
    # Choose a small virtual joint displacement delta_q
    delta_q = np.array([1e-4, -2e-4])
    J = jacobian_2r(q, l1, l2)
    delta_p = J @ delta_q
    deltaW_task = f @ delta_p
    deltaW_joint = tau @ delta_q
    print("deltaW_task ~", deltaW_task, "deltaW_joint ~", deltaW_joint)
      
